core/src/main/scala/io/finch/Output.scala

```diff
@@ -8,37 +8,114 @@ import io.finch.internal.ToResponse
  * An output of [[Endpoint]].
  */
 sealed trait Output[+A] { self =>
+
+  protected def meta: Output.Meta
+  protected def withMeta(meta: Output.Meta): Output[A]
+
+  /**
+   * The status code of this [[Output]].
+   */
+  def status: Status = meta.status
+
+  /**
+   * The header map of this [[Output]].
+   */
+  def headers: Map[String, String] = meta.headers
+
+  /**
+   * The cookie list of this [[Output]].
+   */
+  def cookies: Seq[Cookie] = meta.cookies
+
+  /**
+   * The content type of this [[Output]].
+   */
+  def contentType: Option[String] = meta.contentType
+
+  /**
+   * The charset of this [[Output]].
+   */
+  def charset: Option[String] = meta.charset
+
+  /**
+   * Returns the payload value of this [[Output]] or throws an exception.
+   */
   def value: A
-  def status: Status
-  def headers: Map[String, String]
-  def cookies: Seq[Cookie]
-  def contentType: Option[String]
-  def charset: Option[String]
-
-  def map[B](fn: A => B): Output[B]
-  def flatMap[B](fn: A => Output[B]): Output[B]
-  def flatten[B](implicit ev: A <:< Output[B]): Output[B]
-  def traverse[B](fn: A => Future[B]): Future[Output[B]]
-
-  protected def copy0(
-    headers: Map[String, String] = self.headers,
-    cookies: Seq[Cookie] = self.cookies,
-    contentType: Option[String] = self.contentType,
-    charset: Option[String] = self.charset
-  ): Output[A]
-
-  def withHeader(header: (String, String)): Output[A] = copy0(headers = headers + header)
-  def withCookie(cookie: Cookie): Output[A] = copy0(cookies = cookies :+ cookie)
-  def withContentType(contentType: Option[String]): Output[A] = copy0(contentType = contentType)
-  def withCharset(charset: Option[String]): Output[A] = copy0(charset = charset)
 
+  def map[B](fn: A => B): Output[B] = this match {
+    case Output.Payload(v, m) => Output.Payload(fn(v), m)
+    case f @ Output.Failure(_, _) => f
+    case e @ Output.Empty(_) => e
+  }
+
+  def flatMap[B](fn: A => Output[B]): Output[B] = this match {
+    case p @ Output.Payload(v, _) =>
+      val ob = fn(v)
+      ob.withMeta(ob.meta.copy(
+        headers = ob.headers ++ headers,
+        cookies = ob.cookies ++ cookies,
+        contentType = ob.contentType.orElse(contentType),
+        charset = ob.charset.orElse(charset)
+        )
+      )
+
+    case f @ Output.Failure(_, _) => f
+    case e @ Output.Empty(_) => e
+  }
+
+  def flatten[B](implicit ev: A <:< Output[B]): Output[B] = this match {
+    case p @ Output.Payload(v, _) => v
```

Comment:
I'm not sure how I feel about binding to `p` seeing as it is unused. Maybe this is just all the Elixir I've been doing recently but that screams unused variable to me which is a little bit of a code stink in my mind. I do like how the structure of all the branches looks the same though.